inc/c_util/doublylinkedlist.h

```diff
@@ -4,26 +4,61 @@
 #ifndef DOUBLYLINKEDLIST_H
 #define DOUBLYLINKEDLIST_H
 
-#include "umock_c/umock_c_prod.h"
-
 #ifdef __cplusplus
-extern "C"
-{
+#else
+#include <stdbool.h>
 #endif
 
+#include "macro_utils/macro_utils.h"
+
 typedef struct DLIST_ENTRY_TAG
 {
     struct DLIST_ENTRY_TAG *Flink;
     struct DLIST_ENTRY_TAG *Blink;
 } DLIST_ENTRY, *PDLIST_ENTRY;
 
+#define DLIST_FIND_RESULT_VALUES \
+    DLIST_FIND_RESULT_FOUND, \
+    DLIST_FIND_RESULT_NOT_FOUND, \
+    DLIST_FIND_RESULT_ERROR
+
+MU_DEFINE_ENUM(DLIST_FIND_RESULT, DLIST_FIND_RESULT_VALUES);
+
+#define DLIST_MATCH_FUNCTION_RESULT_VALUES \
+    DLIST_MATCH_FUNCTION_MATCHING, \
+    DLIST_MATCH_FUNCTION_NOT_MATCHING, \
+    DLIST_MATCH_FUNCTION_ERROR
+
+MU_DEFINE_ENUM(DLIST_MATCH_FUNCTION_RESULT, DLIST_MATCH_FUNCTION_RESULT_VALUES);
+
+#define DLIST_CONDITION_FUNCTION_RESULT_VALUES \
+    DLIST_CONDITION_FUNCTION_TRUE, \
+    DLIST_CONDITION_FUNCTION_FALSE, \
+    DLIST_CONDITION_FUNCTION_ERROR
+
+MU_DEFINE_ENUM(DLIST_CONDITION_FUNCTION_RESULT, DLIST_CONDITION_FUNCTION_RESULT_VALUES);
+
+typedef DLIST_MATCH_FUNCTION_RESULT (*DLIST_MATCH_FUNCTION)(PDLIST_ENTRY listEntry, const void* matchContext);
```

Comment:
>const [](http://example.com/codeflow?start=84&length=5)

In general, I don't think EVER we should have a const void* pointer. The reason is that "just" void* should be enough.

To change the memory pointed to by the void*, it *needs* to be cast to something. The moment when it is cast... that's when we should consider const/non const.

Const is just a "pass (or not) the ownership". having const here means "never pass the ownership" - which might be extra requirement as to "don't change anything".

matchContext belongs to the user, so *the user* has ownership on it.

./conisder because we're going to cast the const away every time when it is needed anyway.

Reply:
here and everywhere

Reply:
Agreed. Did it this way because singlylinkedlist had it this way. Removing.